### PR TITLE
Fix draft inputs. Try and fix message overlap

### PR DESCRIPTION
**Pull request by chrisnojima**
Having a hard time trying to reproduce the chat overlap messages. During the process of debugging i saw we're entirely re-creating our chat classes due to our router using unique keys for any route paths (which change when you change conversations). We no longer do this

A side effect of this is the input isn't torn down and recreated each time which means it kept its value (which is a bug anyways as we're supposed to keep the draft values, thats also fixed

I'm clearing the cellCache when convos change (likely doesn't do anything since the whole thing was being blown away, but let's just see). i also updated react-virtualized to latest (shouldn't affect anything)

### CODE REVIEW

File: shared/chat/conversation/input/container.js [hunk 1]
```diff
@@ -82,10 +82,15 @@ export default compose(
         this.props.inputFocus()
       }
     },
+    componentWillUnmount: function () {
+      this.props.onStoreInputText(this.props.inputValue())
```
Comment (chrisnojima) [inline shared/chat/conversation/input/container.js]: mobile unmounts and doesn't really switch convos due to it being on different pages

File: shared/chat/conversation/input/container.js [hunk 2]
```diff
@@ -82,10 +82,15 @@ export default compose(
...
+    },
     componentWillReceiveProps: function (nextProps) {
       if (this.props.selectedConversationIDKey &&
           this.props.selectedConversationIDKey !== nextProps.selectedConversationIDKey) {
         this.props.onStoreInputText(this.props.inputValue())
+        // withState won't get called again if props changes!
+        this.props.setText(nextProps.defaultText)
```
Comment (chrisnojima) [inline shared/chat/conversation/input/container.js]: this happens on desktop when you click between chats

File: shared/chat/conversation/list/index.desktop.js [hunk 3]
```diff
@@ -60,6 +60,7 @@ class BaseList extends Component<void, Props, State> {
   componentWillReceiveProps (nextProps: Props) {
     if (this.props.selectedConversation !== nextProps.selectedConversation ||
       this.props.listScrollDownCounter !== nextProps.listScrollDownCounter) {
+      this._cellCache.clearAll()
```
Comment (chrisnojima) [inline shared/chat/conversation/list/index.desktop.js]: clear the cache when the row to messagkey mapping changes

File: shared/route-tree/render-route.js [hunk 4]
```diff
@@ -113,12 +113,13 @@ function _RenderRoute ({routeDef, routeState, setRouteState, path}: _RenderRoute
...
     stack = childStack
     if (routeDef.containerComponent) {
+      const useKeys = stack.count() > 1 // Only use keys if we have siblings, else we'll render thrash if we keep unique state in here (like in chat)
       // If this route specifies a container component, compose it around every
       // view in the stack.
       stack = stack.map(r => (
         r.update('component', child => (
           <RenderRouteNode
-            key={pathToString(path)}
+            key={useKeys ? pathToString(path) : undefined}
```
Comment (chrisnojima) [inline shared/route-tree/render-route.js]: only use keys if we have siblings. makes chat not rebuild the world while clicking between chats

**Comment by chrisnojima** [comment]
@keybase/react-hackers take a peek

File: shared/route-tree/render-route.js (hunk 4 above)
Comment (chromakode) [inline shared/route-tree/render-route.js]: You can just delete the key here. I don't believe stack items are ever rendered as siblings.

**Comment by chrisnojima** [comment]
@chromakode updated

**Pull request merged**